### PR TITLE
emit ambient error event without crashing

### PR DESCRIPTION
**Pull request by seiyria**
ref: https://forums.tessel.io/t/t2-attempting-to-connect-to-module-causes-process-quit/2325

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -70,9 +70,11 @@ function Ambient(hardware, callback) {
...
 
       // Call the callback with an error
-      if (callback) {
-        callback(err);
-      }
+      setImmediate(function() {
+        if (callback) {
+          callback(err);
+        }
+      });
```
Comment (LinusU) [inline index.js]: It's better to wrap the entire `setImmediate` inside of the `if (callback)`

**Comment by seiyria** [comment]
@LinusU is that what you meant?

**Comment by LinusU** [comment]
Yes 👍, the indentation is wrong now though :)

**Comment by seiyria** [comment]
My bad! Thanks.

**Comment by LinusU** [comment]
Nice 👌 

Actually, you could change that to `setImmediate(callback, err)`

**Comment by seiyria** [comment]
That's true, I was looking for a better way to do it. I'm not nearly as familiar with `setImmediate` as I am with `setInterval` and `setTimeout` so I wanted to play it safe. I'll make that change.

**Comment by LinusU** [comment]
Hehe, sorry to say this now but I don't think that this is the correct fix.

The problem is that `self.attiny.initialize` releases zalgo so it should be fixed in that function...

**Comment by seiyria** [comment]
That's okay. I was going off of the forum post. I'm not sure where initialize is defined, but if you point me there, I'd be happy to close this PR and open a new one with that instead.

**Comment by rwaldron** [comment]
This repo/module needs to have travis setup :\

**Comment by johnnyman727** [comment]
@rwaldron and first it needs tests 😝

**Comment by johnnyman727** [comment]
@seiyria you can find the `initialize` function in the [attiny-common library](https://github.com/tessel/attiny-common/blob/master/lib/index.js#L43). We built that abstraction library because the `ambient` and `ir` module use the same microcontroller so a ton of the code is shared between them.

**Comment by seiyria** [comment]
@johnnyman727 so, should I do what I did here, but [on these lines](https://github.com/tessel/attiny-common/blob/master/lib/index.js#L51-L53)?

**Comment by johnnyman727** [comment]
Actually (sorry to go back and forth here), I'm not sure why @LinusU is suggested you make the change in that library. In seems like the correct thing would be to wrap the `self.emit('error', err)` from within the callback of `self.attiny.initialize`. `self.attiny.initialize` doesn't emit any events - it just calls a callback - so there is no need to wrap that in a setImmediate. It's the call to `use` that can both emit an event and call a callback so I think it's event emissions should be wrapped. @LinusU does that sound correct?

**Comment by LinusU** [comment]
I was suggesting that because I thought that the contract from `self.attiny.initialize` would be to call the callback at a later tick. I might be wrong though, but if that's not the case I think that the correct thing would be to wrap the entire callback in `setImmediate`.

My reasoning is that wrapping just `emit` or just `callback` will leave the same problem on the other path. It's very late here and I've been working for 14 hours straight so I might very well be wrong though :)

**Comment by johnnyman727** [comment]
> My reasoning is that wrapping just emit or just callback will leave the same problem on the other path. 

I'm not sure what you mean here. If you're going the event route, you have to wait for this function to return before you get the object upon which you listen for events (hence why the `setImmediate` is necessary).

Get some sleep!

**Comment by seiyria** [comment]
> I'm not sure what you mean here.

I think he means if we make the fix here, it has to be made in IR too..

**Comment by johnnyman727** [comment]
> I think he means if we make the fix here, it has to be made in IR too.

That makes sense to me. I'm good with the change being made in `attiny-common`. :+1:

**Comment by LinusU** [comment]
Haha :)

I'll try one more time. This is the structure we have:

```javascript
self.attiny.initialize(firmwareOptions, function(err) {
  if (err) {
     self.emit('error', err)
     callback(err)
  }
})
```

And we want to fix it by doing:

```javascript
self.attiny.initialize(firmwareOptions, function(err) {
  if (err) {
     self.emit('error', err)
     setImmediate(callback, err)
  }
})
```

But we also want to do it for the `emit`, so:

```javascript
self.attiny.initialize(firmwareOptions, function(err) {
  if (err) {
     setImmediate(self.emit, 'error', err)
     setImmediate(callback, err)
  }
})
```

But we then we can refactor that to:


```javascript
self.attiny.initialize(firmwareOptions, function(err) {
  setImmediate(function () {
    if (err) {
      self.emit('error', err)
      callback(err)
    }
  })
})
```

But than we we could just fix `initialize` to call the callback in another tick...

-----

That being said, isn't the problem that we emit an error even if a callback handler was added? If so, the fix would be:

```javascript
self.attiny.initialize(firmwareOptions, function(err) {
  if (err) {
    if (callback) {
      callback(err)
    } else {
      self.emit('error', err)
    }
  }
})
```

**Comment by johnnyman727** [comment]
> That being said, isn't the problem that we emit an error even if a callback handler was added? 

Yes, you're totally right per usual. :+1:

**Comment by seiyria** [comment]
So, what should be done here? Should I submit the PR upstream (@LinusU's last snippet of code) or did this get taken care of already?

**Comment by johnnyman727** [comment]
@seiyria if you could add the code from the bottom of @LinusU's comment to this repo, that would be awesome (it just replaces the [top of `attiny.initialize`](https://github.com/tessel/ambient-attx4/blob/master/index.js#L67-L78)).

**Comment by seiyria** [comment]
@johnnyman727 Done, #57.

**Pull request closed without merging**